From 4efd5747f4d20984797732dc64206e7361b79aec Mon Sep 17 00:00:00 2001
From: Josh Fazekas <Josh.Fazekas@Intel.com>
Date: Sat, 07 Jun 2025 02:34:21 +0300
Subject: [PATCH] [SW-230280] gaudi3: prevent HBM access on bad init

Prevent the code from accessing the HBM memory when the first call to
HBM init is successful and a later call fails

Change-Id: Iecda9d2dacd8e0b8f4dbcba4c6eca583c967145a
Signed-off-by: Josh Fazekas <Josh.Fazekas@Intel.com>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index c14baea..4645142 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -387,7 +387,7 @@
 	}
 	total_banks = num_hbms * HBM_PC_NUM * HBM_STACKS_NUM * HBM_BANKS_PER_PC;
 
-	LOG_INF("Performing one time fusescan for HBM to store spare rows in flash");
+	LOG_INF("Performing HBM init row replacement scan");
 	for (hbm_id = 0; hbm_id < num_hbms; hbm_id++) {
 		for (mc_pc = 0; mc_pc < HBM_PC_NUM; mc_pc++) {
 			for (sid = 0; sid < HBM_STACKS_NUM; sid++) {
@@ -521,7 +521,7 @@
 
 __weak void hl_plat_ram_init(void)
 {
-	int err, retry, hbm_mbist_err;
+	int err, retry, hbm_mbist_err, hbm_init_err;
 	uint8_t chosen_hbm_for_mbist;
 	bool fuse_scan_done = false;
 	const struct device *hbm_dev = HBM_DEV_INST;
@@ -555,16 +555,25 @@
 	retry = NUM_HBM_RETRY;
 	while (retry--) {
 		err = preboot_read_hbm_spare_rows_if_empty(hbm_dev, &fuse_scan_done);
-		if (err || fuse_scan_done)
-			err |= gaudi3_hbm_init();
+		if (err)
+			LOG_INF("HBM fuse scan fail");
+		if (err || fuse_scan_done) {
+			hbm_init_err = gaudi3_hbm_init();
+			if (hbm_init_err)
+				LOG_ERR("HBM: init after fuse scan FAILED!");
+			/* Don't differentiate between fuse scan error and init error */
+			err |= hbm_init_err;
+		}
 		if (!err)
 			break;
-		LOG_INF("HBM fuse scan fail; retrying...");
+
+		LOG_INF("Retrying HBM fuse scan");
 	}
 
 	if (err) {
 		LOG_ERR("HBM Full fuse scan failed after multiple retries!!");
 		preboot_set_rma_fuse_scan_fail();
+		cpu_boot_err_set(CPU_BOOT_ERR_DRAM_INIT_FAIL);
 	} else {
 		LOG_INF("HBM: init PASSED!");
 		cpu_boot_sts_set(CPU_BOOT_STATUS_DRAM_RDY);
